python/lsst/dax/apdb/apdbSql.py

```diff
@@ -0,0 +1,889 @@
+        """Return catalog of DiaSource instances from a given region.
+
+        Parameters
+        ----------
+        region : `lsst.sphgeom.Region`
+            Region to search for DIASources.
+        object_ids : iterable [ `int` ], optional
+            List of DiaObject IDs to further constrain the set of returned
+            sources. If `None` then returned sources are not constrained. If
+            list is empty then empty catalog is returned with a correct
+            schema.
+        visit_time : `lsst.daf.base.DateTime`
+            Time of the current visit.
+
+        Returns
+        -------
+        catalog : `pandas.DataFrame`, or `None`
+            Catalog containing DiaSource records. `None` is returned if
+            ``read_sources_months`` configuration parameter is set to 0.
+
+        Notes
+        -----
+        This method returns DiaSource catalog for a region with additional
+        filtering based on DiaObject IDs. Only as subset of DiaSource history
+        is returned limited by ``read_sources_months`` config parameter, w.r.t.
+        ``visit_time``. If ``object_ids`` is empty then an empty catalog is
+        always returned with a correct schema (columns/types). If
+        ``object_ids`` is `None` then no filtering is performed and some of the
+        returned records may be outside the specified region.
+        """
+        if self.config.read_sources_months == 0:
+            _LOG.debug("Skip DiaSources fetching")
+            return None
+
+        if object_ids is None:
+            # region-based select
+            return self._getDiaSourcesInRegion(region, visit_time)
+        else:
+            return self._getDiaSourcesByIDs(list(object_ids), visit_time)
+
+    def getDiaForcedSources(self, region: Region,
+                            object_ids: Optional[Iterable[int]],
+                            visit_time: dafBase.DateTime) -> Optional[pandas.DataFrame]:
+        """Return catalog of DiaForcedSource instances from a given region.
+
+        Parameters
+        ----------
+        region : `lsst.sphgeom.Region`
+            Region to search for DIASources.
+        object_ids : iterable [ `int` ], optional
+            List of DiaObject IDs to further constrain the set of returned
+            sources. If list is empty then empty catalog is returned with a
+            correct schema. If `None` then returned sources are not
+            constrained. Some implementations may not support latter case.
+        visit_time : `lsst.daf.base.DateTime`
+            Time of the current visit.
+
+        Returns
+        -------
+        catalog : `pandas.DataFrame`, or `None`
+            Catalog containing DiaSource records. `None` is returned if
+            ``read_sources_months`` configuration parameter is set to 0.
+
+        Raises
+        ------
+        NotImplementedError
+            May be raised by some implementations if ``object_ids`` is `None`.
```

Comment:
I should have dropped docstrings for inherited methods in this class, I removed them in later commit, will move those changes to this commit. This docstring is just a copy from the base class. `ApdbSql` is a fully implemented (it is final now but you can still extend it if you want). `ApdbSql` does raise if `object_ids` is `None`, but future Cassandra-based implementation can work OK without `object_ids`. I thought I would keep this option available even if it works differently across implementations, could be a chance to optimize things based on backend.

Reply:
I meant that if the `AbdpSql` version always requires `object_ids`, perhaps it should document this as an absolute statement rather than a conditional one.

Reply:
Got it, I changed the docstring to be more specific for this class (and different from base class).